File: packages/sdk-python/agentguard/redact.py
"""Apply the runtime's redaction paths to a parameter dict."""

from __future__ import annotations

import copy
import re

REDACTED = "[REDACTED]"
_INDEX = re.compile(r"\[(\d+)\]")


def _tokens(path: str) -> list[str | int]:
    # "parameters.user.contacts[0].email" -> ["user", "contacts", 0, "email"]
    body = path.split(".", 1)[1] if path.startswith("parameters.") else path
    out: list[str | int] = []
    for part in body.split("."):
        m = _INDEX.search(part)
        name = _INDEX.sub("", part)
        if name:
            out.append(name)
        if m:
            out.append(int(m.group(1)))
    return out


def redact_params(parameters: dict, paths: list[str]) -> dict:
    """Return a deep copy of ``parameters`` with every ``path`` leaf masked."""
    result = copy.deepcopy(parameters)
    for path in paths:
        tokens = _tokens(path)
        if not tokens:
            continue
        node = result
        try:
            for key in tokens[:-1]:
                node = node[key]
            node[tokens[-1]] = REDACTED
        except (KeyError, IndexError, TypeError):
            continue
    return result

File: packages/sdk-typescript/src/redact.ts
/** Apply the runtime's redaction paths to a parameter object. */

export const REDACTED = "[REDACTED]";

type Key = string | number;

/** "parameters.user.contacts[0].email" -> ["user", "contacts", 0, "email"] */
function tokens(path: string): Key[] {
  const body = path.startsWith("parameters.") ? path.slice("parameters.".length) : path;
  const out: Key[] = [];
  for (const part of body.split(".")) {
    const match = /\[(\d+)\]/.exec(part);
    const name = part.replace(/\[\d+\]/g, "");
    if (name) out.push(name);
    if (match) out.push(Number(match[1]));
  }
  return out;
}

function clone<T>(value: T): T {
  return typeof structuredClone === "function"
    ? structuredClone(value)
    : (JSON.parse(JSON.stringify(value)) as T);
}

/** Return a deep copy of `parameters` with every `path` leaf masked. */
export function applyRedactions(
  parameters: Record<string, unknown>,
  paths: string[],
): Record<string, unknown> {
  const result = clone(parameters);
  for (const path of paths) {
    const parts = tokens(path);
    if (parts.length === 0) continue;
    let node: unknown = result;
    let ok = true;
    for (const key of parts.slice(0, -1)) {
      if (node != null && typeof node === "object") {
        node = (node as Record<Key, unknown>)[key];
      } else {
        ok = false;
        break;
      }
    }
    const last = parts[parts.length - 1]!;
    if (ok && node != null && typeof node === "object") {
      (node as Record<Key, unknown>)[last] = REDACTED;
    }
  }
  return result;
}
